keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -164,100 +165,105 @@ that is runtime agnostic, and will contain platform-specific information.
   updated, it updates Pod.Status.ContainerStatuses[i].Resources to reflect the
   new resource values.
-* If new desired resources doesn't fit, Kubelet will reject the resize, and no
-  further changes are made.
-  - Kubelet retries Pod resize at a later time, or when other Pods depart and
-    free up resources.
+* If new desired resources don't fit, Kubelet rejects the resize, and no
+  further action is taken.
+  - Kubelet retries the Pod resize at a later time.
+
+Scheduler may, in parallel, assign a new Pod to the Node because it uses cached
+Pods to compute Node allocable values. If this race condition occurs, Kubelet
+resolves it by rejecting that new Pod if the Node has no room after Pod resize.
 
-Kubelet uses max(Pod.Spec.Containers[i].Resources,
-Pod.Status.ContainerStatuses[i].Resources) for computing Node resource usage
-to avoid race between competing Pod resize requests.
+#### Kubelet Restart Tolerance
 
-Scheduler may, in parallel, assign a new Pod to the Node because it uses
-cached Node resources values. By using max(Pod.Spec.Containers[i].Resources,
-Pod.Status.ContainerStatuses[i].Resources) Kubelet also prevents new Pods
-from competing with Pod resize, and rejects a new Pod if Node does not have
-enough room.
+If Kubelet were to restart amidst handling a Pod resize, then upon restart, all
+existing Pods, new & departed Pods, and Pods needing resize are handled as follows:
 
-Additionally, Kubelet may evict lower priority Pods from the Node in order to
-make room for the resize. Eviction of lower priority Pods can be done in
-second phase of the implementation of this feature. (not scoped for this KEP)
+1. All existing Pods (including Pods needing resources resize) are admitted at
+   their current Pod.Spec.Containers[i].ResourcesAllocated values, with Pods
+   sorted by their creation time.
+1. All departed Pods are cleaned up.
+1. Once all existing (and departed) Pods are accounted for, new Pods & Pods
+   needing resize (i.e Pods for which Spec.Containers[i].Resources.Requests
+   and Spec.Containers[i].ResourcesAllocated differ) are sorted by increasing
+   ResourceVersion, and admitted as long as Node has enough room.
+
+This will ensure first-come-first-serve processing, just as they would have
+been handled if Kubelet had never restarted.
 
 ### Scheduler and API Server Interaction
 
-Scheduler observes the resize request posted to API Server, and updates the
-Node available resources accounting in its cache by using Pod's
-max(Spec.Containers[i].Resources, Status.ContainerStatuses[i].Resources) when
-computing Node resources used by the Pods. This ensures that, in the case of
-resource decrease for existing Pod, new Pod is not prematurely assigned to the
-Node that is still in the process of deallocating the resized Pod's resources.
+Scheduler continues to watch changes to Pods and updates its cache. It uses the
+cached Pod's Spec.Containers[i].ResourcesAllocated values to compute the Node
```

Comment:
This "summation" of container resources to calculate pod resources needs to take Pod Overhead into account (this is at pod level) to get an accurate view of the actual resources associated with the pod.

Reply:
Yes. I've captured this in the [Notes](https://github.com/kubernetes/enhancements/pull/686/files#diff-57e345a188df5653422eb0ed20bf9442R291) section of the KEP. It is a constant that gets taken into account when checking for it. 

I pulled down latest master and don't see it applied in Kubelet. Shouldn't Kubelet's admission control check for this? (multiple scheduler race)

CC: @tallclair

Reply:
If Kubelet is calling scheduler predicates and it accounts for Pod overhead, then our admission check should be covered as well. If not, and we find issues during testing, we will fix it. Resolving this issue.